core/vm/contracts.go

```diff
@@ -715,10 +753,99 @@ func (c *hashHeader) Run(input []byte, caller common.Address, evm *EVM, gas uint
 	}
 
 	hashBytes := header.Hash().Bytes()
 
 	return hashBytes, gas, nil
 }
+
+type getParentSealBitmap struct{}
+
+func (c *getParentSealBitmap) RequiredGas(input []byte) uint64 {
+	return params.GetParentSealBitmapGas
+}
+
+// Return the signer bitmap from the parent seal of a past block in the chain.
+// Requested parent seal must have occurred within 4 epochs of the current block number.
+func (c *getParentSealBitmap) Run(input []byte, caller common.Address, evm *EVM, gas uint64) ([]byte, uint64, error) {
+	gas, err := debitRequiredGas(c, input, gas)
+	if err != nil {
+		return nil, gas, err
+	}
+
+	// input is comprised of a single argument:
+	//   blockNumber: 32 byte integer representing the block number to access
+	if len(input) < 32 {
+		return nil, gas, ErrInputLength
+	}
+
+	blockNumber := new(big.Int).SetBytes(input[0:32])
+
+	// Ensure the request is for information from a previously sealed block.
+	if blockNumber.Cmp(common.Big0) == 0 {
+		return nil, gas, ErrBlockNumberOutOfBounds
+	}
+	if blockNumber.Cmp(evm.Context.BlockNumber) >= 0 {
+		return nil, gas, ErrBlockNumberOutOfBounds
+	}
+
+	// Ensure the request is for a sufficiently recent block to limit state expansion.
+	historyLimit := new(big.Int).SetUint64(evm.Context.Engine.EpochSize() * 4)
+	if blockNumber.Cmp(new(big.Int).Sub(evm.Context.BlockNumber, historyLimit)) <= 0 {
+		return nil, gas, ErrBlockNumberOutOfBounds
+	}
+
+	header := evm.Context.GetHeaderByNumber(blockNumber.Uint64())
+	if header == nil {
+		// Under sane circumstances, this error should never be reached.
+		return nil, gas, ErrUnexpected
+	}
+
+	extra, err := types.ExtractIstanbulExtra(header)
+	if err != nil {
+		// Header is from a non-Istanbul engine.
+		return nil, gas, ErrEngineIncompatible
+	}
+
+	return common.LeftPadBytes(extra.ParentAggregatedSeal.Bitmap.Bytes()[:], 32), gas, nil
+}
+
+// getVerifiedSealBitmap is a precompile to verify the seal on a given header and extract its bitmap.
+type getVerifiedSealBitmap struct{}
+
+func (c *getVerifiedSealBitmap) RequiredGas(input []byte) uint64 {
+	return params.GetVerifiedSealBitmapGas
+}
+
+func (c *getVerifiedSealBitmap) Run(input []byte, caller common.Address, evm *EVM, gas uint64) ([]byte, uint64, error) {
+	gas, err := debitRequiredGas(c, input, gas)
+	if err != nil {
+		return nil, gas, err
+	}
+
+	// input is comprised of a single argument:
+	//   header:  rlp encoded block header
+	if len(input) == 0 {
+		return nil, gas, ErrInputLength
+	}
+
+	// Decode and verify the seal against the engine rules.
+	header := new(types.Header)
+	if err := rlp.DecodeBytes(input, header); err != nil {
+		return nil, gas, ErrInputDecode
+	}
+	if !evm.Context.VerifySeal(header) {
+		return nil, gas, ErrInputValidation
+	}
+
+	// Extract the verified seal from the header.
+	extra, err := types.ExtractIstanbulExtra(header)
+	if err != nil {
+		// Seal verified by a non-Istanbul engine. Return an error.
+		return nil, gas, ErrEngineIncompatible
```

Comment:
Unlike the previous precompile, this is possible to reach since the user can put any data they want. They should maybe be different errors

Reply:
It's not possible to hit as long as we are using the Istanbul engine because seal verification uses the Istanbul extra data (namely the aggregated seal)

Reply:
But a user could pass invalid istanbul extradata, right?